[<Fury component: Java>] Example in readme has runtime exception

**Describe the bug**

Code example of Cross-language object graph serialization in Java has Runtime exception.

https://github.com/alipay/fury/blob/main/README.md?plain=1#L176

**Expected behavior**

Complete without exception.

**To Reproduce**

mvn exec:java -Dexec.mainClass="io.fury.examples.ReferenceExample"

**Screenshots**

[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.1.0:java (default-cli) on project fury: An exception occurred while executing the Java class. class com.google.common.collect.RegularImmutableMap is not registered, if it's not the type you want to serialize, it may be a **vulnerability**. If it's not a vulnerability, registering class by `Fury#register` will have better performance, otherwise class name will be serialized too. -> [Help 1]

**Environment (please complete the following information):**

openjdk 17.0.7 

**Fix**

 `
fury.register(Class.forName("com.google.common.collect.RegularImmutableMap"),"com.google.common.collect.RegularImmutableMap");
`

--- a/java/fury-core/src/main/java/io/fury/Config.java
+++ b/java/fury-core/src/main/java/io/fury/Config.java
@@ -44,6 +44,7 @@ public class Config implements Serializable {
   private final boolean compressNumber;
   private final boolean secureModeEnabled;
   private final boolean requireClassRegistration;
+  private final boolean registerGuavaTypes;
   private final boolean metaContextShareEnabled;
   private final boolean asyncCompilationEnabled;
   private final boolean deserializeUnExistClassEnabled;
@@ -59,6 +60,7 @@ public class Config implements Serializable {
     compressNumber = builder.compressNumber;
     secureModeEnabled = builder.secureModeEnabled;
     requireClassRegistration = builder.requireClassRegistration;
+    registerGuavaTypes = builder.registerGuavaTypes;
     codeGenEnabled = builder.codeGenEnabled;
     checkClassVersion = builder.checkClassVersion;
     compatibleMode = builder.compatibleMode;
@@ -131,6 +133,10 @@ public class Config implements Serializable {
 
   public boolean requireClassRegistration() {
     return requireClassRegistration;
+  }
+
+  public boolean registerGuavaTypes() {
+    return registerGuavaTypes;
   }
 
   /**

--- a/java/fury-core/src/main/java/io/fury/Fury.java
+++ b/java/fury-core/src/main/java/io/fury/Fury.java
@@ -1279,6 +1279,7 @@ public final class Fury {
     boolean codeGenEnabled = true;
     public boolean deserializeUnExistClassEnabled = false;
     public boolean asyncCompilationEnabled = false;
+    public boolean registerGuavaTypes = true;
 
     private FuryBuilder() {}
 
@@ -1353,6 +1354,17 @@ public final class Fury {
     public FuryBuilder withDefaultJDKCompatibleSerializerType(
         Class<? extends Serializer> serializerClass) {
       this.defaultJDKStreamSerializerType = serializerClass;
+      return this;
+    }
+
+    /**
+     * Whether pre-register guava types such as `RegularImmutableMap`/`RegularImmutableList`. Those
+     * types are not public API, but seems pretty stable.
+     *
+     * @see io.fury.serializer.GuavaSerializers
+     */
+    public FuryBuilder registerGuavaTypes(boolean register) {
+      this.registerGuavaTypes = register;
       return this;
     }
 

--- a/java/fury-core/src/main/java/io/fury/resolver/ClassResolver.java
+++ b/java/fury-core/src/main/java/io/fury/resolver/ClassResolver.java
@@ -57,6 +57,7 @@ import io.fury.serializer.CollectionSerializers;
 import io.fury.serializer.CompatibleMode;
 import io.fury.serializer.CompatibleSerializer;
 import io.fury.serializer.ExternalizableSerializer;
+import io.fury.serializer.GuavaSerializers;
 import io.fury.serializer.JavaSerializer;
 import io.fury.serializer.JdkProxySerializer;
 import io.fury.serializer.LambdaSerializer;
@@ -322,6 +323,9 @@ public class ClassResolver {
     if (metaContextShareEnabled) {
       addDefaultSerializer(
           UnexistedMetaSharedClass.class, new UnexistedClassSerializer(fury, null));
+    }
+    if (fury.getConfig().registerGuavaTypes()) {
+      GuavaSerializers.registerDefaultSerializers(fury);
     }
     if (ArrowSerializersClass != null) {
       try {
@@ -712,9 +716,9 @@ public class ClassResolver {
       } else if (Externalizable.class.isAssignableFrom(cls)) {
         return ExternalizableSerializer.class;
       } else if (ImmutableList.class.isAssignableFrom(cls)) {
-        return CollectionSerializers.ImmutableListSerializer.class;
+        return GuavaSerializers.ImmutableListSerializer.class;
       } else if (ImmutableMap.class.isAssignableFrom(cls)) {
-        return MapSerializers.ImmutableMapSerializer.class;
+        return GuavaSerializers.ImmutableMapSerializer.class;
       } else if (ByteBuffer.class.isAssignableFrom(cls)) {
         return BufferSerializers.ByteBufferSerializer.class;
       }
@@ -755,6 +759,9 @@ public class ClassResolver {
         } else {
           return MapSerializers.MapSerializer.class;
         }
+      }
+      if (fury.getLanguage() != Language.JAVA) {
+        LOG.warn("Class {} isn't supported for cross-language serialization.", cls);
       }
       if (useReplaceResolveSerializer(cls)) {
         return ReplaceResolveSerializer.class;
@@ -797,9 +804,6 @@ public class ClassResolver {
       boolean shareMeta,
       boolean codegen,
       JITContext.SerializerJITCallback<Class<? extends Serializer>> callback) {
-    if (fury.getLanguage() != Language.JAVA) {
-      LOG.warn("Class {} isn't supported for cross-language serialization.", cls);
-    }
     if (codegen) {
       if (extRegistry.getClassCtx.contains(cls)) {
         // avoid potential recursive call for seq codec generation.

--- a/java/fury-core/src/main/java/io/fury/serializer/CollectionSerializers.java
+++ b/java/fury-core/src/main/java/io/fury/serializer/CollectionSerializers.java
@@ -21,7 +21,6 @@ package io.fury.serializer;
 import static io.fury.type.TypeUtils.getRawType;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.ImmutableList;
 import com.google.common.reflect.TypeToken;
 import io.fury.Fury;
 import io.fury.Language;
@@ -698,43 +697,6 @@ public class CollectionSerializers {
     public Set<?> xread(MemoryBuffer buffer) {
       buffer.readPositiveVarInt();
       return Collections.singleton(fury.xreadRef(buffer));
-    }
-  }
-
-  public static final class ImmutableListSerializer<T extends ImmutableList>
-      extends CollectionSerializer<T> {
-    private final ReplaceResolveSerializer serializer;
-
-    public ImmutableListSerializer(Fury fury, Class<T> cls) {
-      super(fury, cls, false, false);
-      fury.getClassResolver().setSerializer(cls, this);
-      serializer = new ReplaceResolveSerializer(fury, cls);
-    }
-
-    @SuppressWarnings("unchecked")
-    @Override
-    public T read(MemoryBuffer buffer) {
-      return (T) serializer.read(buffer);
-    }
-
-    @Override
-    public void write(MemoryBuffer buffer, T value) {
-      serializer.write(buffer, value);
-    }
-
-    @Override
-    public short getXtypeId() {
-      return (short) -Type.LIST.getId();
-    }
-
-    @Override
-    public T xread(MemoryBuffer buffer) {
-      int size = buffer.readPositiveVarInt();
-      List list = new ArrayList<>();
-      xreadElements(fury, buffer, list, size);
-      T immutableList = (T) ImmutableList.copyOf(list);
-      fury.getRefResolver().reference(immutableList);
-      return immutableList;
     }
   }
 

--- a/java/fury-core/src/main/java/io/fury/serializer/GuavaSerializers.java
+++ b/java/fury-core/src/main/java/io/fury/serializer/GuavaSerializers.java
@@ -18,7 +18,173 @@
 
 package io.fury.serializer;
 
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.fury.Fury;
+import io.fury.memory.MemoryBuffer;
+import io.fury.serializer.CollectionSerializers.CollectionSerializer;
+import io.fury.serializer.MapSerializers.MapSerializer;
+import io.fury.type.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Serializers for common guava types.
+ *
+ * @author chaokunyang
+ */
+@SuppressWarnings({"unchecked", "rawtypes"})
 public class GuavaSerializers {
+  // TODO(chaokunyang) support jit optimization for guava types.
+
+  abstract static class GuavaCollectionSerializer<T extends Collection>
+      extends CollectionSerializer<T> {
+    private final ReplaceResolveSerializer serializer;
+
+    public GuavaCollectionSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls, false, false);
+      fury.getClassResolver().setSerializer(cls, this);
+      serializer = new ReplaceResolveSerializer(fury, cls);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T read(MemoryBuffer buffer) {
+      return (T) serializer.read(buffer);
+    }
+
+    @Override
+    public void write(MemoryBuffer buffer, T value) {
+      serializer.write(buffer, value);
+    }
+
+    @Override
+    public T xread(MemoryBuffer buffer) {
+      int size = buffer.readPositiveVarInt();
+      List list = new ArrayList<>();
+      xreadElements(fury, buffer, list, size);
+      T immutableList = xnewInstance(list);
+      fury.getRefResolver().reference(immutableList);
+      return immutableList;
+    }
+
+    protected abstract T xnewInstance(Collection collection);
+  }
+
+  public static final class ImmutableListSerializer<T extends ImmutableList>
+      extends GuavaCollectionSerializer<T> {
+    public ImmutableListSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls);
+      fury.getClassResolver().setSerializer(cls, this);
+    }
+
+    @Override
+    public short getXtypeId() {
+      return (short) -Type.LIST.getId();
+    }
+
+    @Override
+    public T xread(MemoryBuffer buffer) {
+      int size = buffer.readPositiveVarInt();
+      List list = new ArrayList<>();
+      xreadElements(fury, buffer, list, size);
+      T immutableList = (T) ImmutableList.copyOf(list);
+      fury.getRefResolver().reference(immutableList);
+      return immutableList;
+    }
+
+    public T xnewInstance(Collection collection) {
+      return (T) ImmutableList.copyOf(collection);
+    }
+  }
+
+  public static final class ImmutableSetSerializer<T extends ImmutableSet>
+      extends GuavaCollectionSerializer<T> {
+
+    public ImmutableSetSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls);
+    }
+
+    @Override
+    public short getXtypeId() {
+      return (short) -Type.FURY_SET.getId();
+    }
+
+    @Override
+    protected T xnewInstance(Collection collection) {
+      return (T) ImmutableSet.copyOf(collection);
+    }
+  }
+
+  abstract static class GuavaMapSerializer<T extends Map> extends MapSerializer<T> {
+    private final ReplaceResolveSerializer serializer;
+
+    public GuavaMapSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls, false, false);
+      fury.getClassResolver().setSerializer(cls, this);
+      serializer = new ReplaceResolveSerializer(fury, cls);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T read(MemoryBuffer buffer) {
+      return (T) serializer.read(buffer);
+    }
+
+    @Override
+    public void write(MemoryBuffer buffer, T value) {
+      serializer.write(buffer, value);
+    }
+
+    @Override
+    public short getXtypeId() {
+      return (short) -Type.MAP.getId();
+    }
+
+    @Override
+    public T xread(MemoryBuffer buffer) {
+      int size = buffer.readPositiveVarInt();
+      Map map = new HashMap();
+      xreadElements(fury, buffer, map, size);
+      T immutableMap = xnewInstance(map);
+      fury.getRefResolver().reference(immutableMap);
+      return immutableMap;
+    }
+
+    protected abstract T xnewInstance(Map map);
+  }
+
+  public static final class ImmutableMapSerializer<T extends ImmutableMap>
+      extends GuavaMapSerializer<T> {
+
+    public ImmutableMapSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls);
+      fury.getClassResolver().setSerializer(cls, this);
+    }
+
+    @Override
+    protected T xnewInstance(Map map) {
+      return (T) ImmutableMap.copyOf(map);
+    }
+  }
+
+  public static final class ImmutableBiMapSerializer<T extends ImmutableBiMap>
+      extends GuavaMapSerializer<T> {
+    public ImmutableBiMapSerializer(Fury fury, Class<T> cls) {
+      super(fury, cls);
+      fury.getClassResolver().setSerializer(cls, this);
+    }
+
+    @Override
+    protected T xnewInstance(Map map) {
+      return (T) ImmutableBiMap.copyOf(map);
+    }
+  }
 
   // TODO guava serializers
   // guava/ArrayListMultimapSerializer - serializer for guava-libraries' ArrayListMultimap
@@ -38,5 +204,44 @@ public class GuavaSerializers {
   // guava/TreeMultimapSerializer - serializer for guava-libraries' TreeMultimap
   // guava/UnmodifiableNavigableSetSerializer - serializer for guava-libraries'
   // UnmodifiableNavigableSet
+  public static void registerDefaultSerializers(Fury fury) {
+    // Note: Guava common types are not public API, don't register by `ImmutableXXX.of()`,
+    // since different guava version may return different type objects, which make class
+    // registration
+    // inconsistent if peers load different version of guava.
+    // For example: guava 20 return ImmutableBiMap for ImmutableMap.of(), but guava 27 return
+    // ImmutableMap.
+    String pkg = "com.google.common.collect";
+    Class<?> cls = loadClass(pkg + ".RegularImmutableBiMap", ImmutableBiMap.of().getClass());
+    fury.registerSerializer(cls, new ImmutableBiMapSerializer(fury, cls));
+    cls = loadClass(pkg + ".SingletonImmutableBiMap", ImmutableBiMap.of(1, 2).getClass());
+    fury.registerSerializer(cls, new ImmutableBiMapSerializer(fury, cls));
+    cls = loadClass(pkg + ".RegularImmutableMap", ImmutableMap.of().getClass());
+    fury.registerSerializer(cls, new ImmutableMapSerializer(fury, cls));
+    cls = loadClass(pkg + ".RegularImmutableList", ImmutableList.of().getClass());
+    fury.registerSerializer(cls, new ImmutableListSerializer(fury, cls));
+    cls = loadClass(pkg + ".SingletonImmutableList", ImmutableList.of(1).getClass());
+    fury.registerSerializer(cls, new ImmutableSetSerializer(fury, cls));
+    cls = loadClass(pkg + ".RegularImmutableSet", ImmutableSet.of().getClass());
+    fury.registerSerializer(cls, new ImmutableSetSerializer(fury, cls));
+    cls = loadClass(pkg + ".SingletonImmutableSet", ImmutableSet.of(1).getClass());
+    fury.registerSerializer(cls, new ImmutableSetSerializer(fury, cls));
+    // TODO(chaokunyang) add immutable sorted set support.
+    //  sorted set doesn't support xlang.
+    // register(fury, loadClass(pkg + ".RegularImmutableSortedSet",
+    // ImmutableSortedSet.of(1).getClass()));
+    // register(fury, loadClass(pkg + ".ImmutableSortedMap", ImmutableSortedMap.of().getClass()));
+  }
 
+  static Class<?> loadClass(String className, Class<?> cache) {
+    if (cache.getName().equals(className)) {
+      return cache;
+    } else {
+      try {
+        return Class.forName(className);
+      } catch (ClassNotFoundException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
 }

--- a/java/fury-core/src/main/java/io/fury/serializer/MapSerializers.java
+++ b/java/fury-core/src/main/java/io/fury/serializer/MapSerializers.java
@@ -22,7 +22,6 @@ import static io.fury.type.TypeUtils.MAP_TYPE;
 import static io.fury.type.TypeUtils.getRawType;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.ImmutableMap;
 import com.google.common.reflect.TypeToken;
 import io.fury.Fury;
 import io.fury.Language;
@@ -60,7 +59,7 @@ import java.util.concurrent.ConcurrentSkipListMap;
  *
  * @author chaokunyang
  */
-@SuppressWarnings({"unchecked", "rawtypes", "UnstableApiUsage"})
+@SuppressWarnings({"unchecked", "rawtypes"})
 public class MapSerializers {
 
   public static class MapSerializer<T extends Map> extends Serializer<T> {
@@ -968,43 +967,6 @@ public class MapSerializers {
       Object key = fury.xreadRef(buffer);
       Object value = fury.xreadRef(buffer);
       return Collections.singletonMap(key, value);
-    }
-  }
-
-  public static final class ImmutableMapSerializer<T extends ImmutableMap>
-      extends MapSerializer<T> {
-    private final ReplaceResolveSerializer serializer;
-
-    public ImmutableMapSerializer(Fury fury, Class<T> cls) {
-      super(fury, cls, false, false);
-      fury.getClassResolver().setSerializer(cls, this);
-      serializer = new ReplaceResolveSerializer(fury, cls);
-    }
-
-    @SuppressWarnings("unchecked")
-    @Override
-    public T read(MemoryBuffer buffer) {
-      return (T) serializer.read(buffer);
-    }
-
-    @Override
-    public void write(MemoryBuffer buffer, T value) {
-      serializer.write(buffer, value);
-    }
-
-    @Override
-    public short getXtypeId() {
-      return (short) -Type.MAP.getId();
-    }
-
-    @Override
-    public T xread(MemoryBuffer buffer) {
-      int size = buffer.readPositiveVarInt();
-      Map map = new HashMap();
-      xreadElements(fury, buffer, map, size);
-      T immutableMap = (T) ImmutableMap.copyOf(map);
-      fury.getRefResolver().reference(immutableMap);
-      return immutableMap;
     }
   }
 

--- a/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
@@ -24,12 +24,14 @@ import static org.testng.Assert.assertSame;
 import com.google.common.base.Objects;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
 import com.google.common.reflect.TypeToken;
 import io.fury.Fury;
 import io.fury.FuryTestBase;
 import io.fury.Language;
 import io.fury.memory.MemoryBuffer;
 import io.fury.memory.MemoryUtils;
+import io.fury.serializer.CollectionSerializers.JDKCompatibleCollectionSerializer;
 import io.fury.type.GenericType;
 import java.io.Serializable;
 import java.util.AbstractCollection;
@@ -388,8 +390,8 @@ public class CollectionSerializersTest extends FuryTestBase {
   @SuppressWarnings("unchecked")
   @Test
   public void testJavaSerialization() {
-    ImmutableSet<Integer> set = ImmutableSet.of(1, 2, 3);
-    Class<? extends ImmutableSet> setClass = set.getClass();
+    ImmutableSortedSet<Integer> set = ImmutableSortedSet.of(1, 2, 3);
+    Class<? extends ImmutableSortedSet> setClass = set.getClass();
     Fury fury =
         Fury.builder()
             .withLanguage(Language.JAVA)
@@ -397,8 +399,7 @@ public class CollectionSerializersTest extends FuryTestBase {
             .disableSecureMode()
             .build();
     MemoryBuffer buffer = MemoryUtils.buffer(32);
-    CollectionSerializers.JDKCompatibleCollectionSerializer javaSerializer =
-        new CollectionSerializers.JDKCompatibleCollectionSerializer(fury, setClass);
+    JDKCompatibleCollectionSerializer javaSerializer = new JDKCompatibleCollectionSerializer(fury, setClass);
     javaSerializer.write(buffer, set);
     Object read = javaSerializer.read(buffer);
     assertEquals(set, read);

--- a/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
@@ -399,7 +399,8 @@ public class CollectionSerializersTest extends FuryTestBase {
             .disableSecureMode()
             .build();
     MemoryBuffer buffer = MemoryUtils.buffer(32);
-    JDKCompatibleCollectionSerializer javaSerializer = new JDKCompatibleCollectionSerializer(fury, setClass);
+    JDKCompatibleCollectionSerializer javaSerializer =
+        new JDKCompatibleCollectionSerializer(fury, setClass);
     javaSerializer.write(buffer, set);
     Object read = javaSerializer.read(buffer);
     assertEquals(set, read);

--- a/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
+++ b/java/fury-core/src/test/java/io/fury/serializer/CollectionSerializersTest.java
@@ -212,14 +212,6 @@ public class CollectionSerializersTest extends FuryTestBase {
   }
 
   @Test
-  public void testImmutableListSerializer() {
-    serDe(getJavaFury(), ImmutableList.of(1, 2));
-    Assert.assertEquals(
-        getJavaFury().getClassResolver().getSerializerClass(ImmutableList.of(1, 2).getClass()),
-        CollectionSerializers.ImmutableListSerializer.class);
-  }
-
-  @Test
   public void testSerializeJavaBlockingQueue() {
     Fury fury =
         Fury.builder()

--- a/java/fury-core/src/test/java/io/fury/serializer/GuavaSerializersTest.java
+++ b/java/fury-core/src/test/java/io/fury/serializer/GuavaSerializersTest.java
@@ -1,0 +1,68 @@
+/*
+ * Copyright 2023 The Fury authors
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.fury.serializer;
+
+import static org.testng.Assert.*;
+
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.fury.Fury;
+import io.fury.FuryTestBase;
+import io.fury.Language;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class GuavaSerializersTest extends FuryTestBase {
+
+  @Test
+  public void testImmutableListSerializer() {
+    serDe(getJavaFury(), ImmutableList.of(1, 2));
+    Assert.assertEquals(
+        getJavaFury().getClassResolver().getSerializerClass(ImmutableList.of(1, 2).getClass()),
+        GuavaSerializers.ImmutableListSerializer.class);
+  }
+
+  @Test
+  public void testImmutableMapSerializer() {
+    serDe(getJavaFury(), ImmutableMap.of("k1", 1, "k2", 2));
+    Assert.assertEquals(
+        getJavaFury()
+            .getClassResolver()
+            .getSerializerClass(ImmutableMap.of("k1", 1, "k2", 2).getClass()),
+        GuavaSerializers.ImmutableMapSerializer.class);
+  }
+
+  @Test
+  public void tesXlangSerialize() {
+    Fury fury = Fury.builder().withLanguage(Language.XLANG).build();
+    serDe(fury, ImmutableBiMap.of());
+    serDe(fury, ImmutableBiMap.of(1, 2));
+    serDe(fury, ImmutableBiMap.of(1, 2, 3, 4));
+
+    serDe(fury, ImmutableList.of());
+    serDe(fury, ImmutableList.of(1));
+    serDe(fury, ImmutableList.of(1, 2));
+
+    serDe(fury, ImmutableSet.of());
+    serDe(fury, ImmutableSet.of(1));
+    serDe(fury, ImmutableSet.of(1, 2, 3, 4));
+  }
+}

--- a/java/fury-core/src/test/java/io/fury/serializer/MapSerializersTest.java
+++ b/java/fury-core/src/test/java/io/fury/serializer/MapSerializersTest.java
@@ -166,16 +166,6 @@ public class MapSerializersTest extends FuryTestBase {
         MapSerializers.EnumMapSerializer.class);
   }
 
-  @Test
-  public void testImmutableMapSerializer() {
-    serDe(getJavaFury(), ImmutableMap.of("k1", 1, "k2", 2));
-    Assert.assertEquals(
-        getJavaFury()
-            .getClassResolver()
-            .getSerializerClass(ImmutableMap.of("k1", 1, "k2", 2).getClass()),
-        MapSerializers.ImmutableMapSerializer.class);
-  }
-
   private static Map<String, Integer> newInnerMap() {
     return new HashMap<String, Integer>() {
       {

